File: 0101-0200/0174.Dungeon_Game.py
#7ms





from typing import List

class Solution:
    def calculateMinimumHP(self, dungeon: List[List[int]]) -> int:

        rows, cols = len(dungeon), len(dungeon[0])

        dp = [[0] * cols for _ in range(rows)]
        dp[-1][-1] = max(1, 1 - dungeon[-1][-1])

        for i in range(rows - 2, -1, -1):
            dp[i][-1] = max(1, dp[i + 1][-1] - dungeon[i][-1])

        for j in range(cols - 2, -1, -1):
            dp[-1][j] = max(1, dp[-1][j + 1] - dungeon[-1][j])

        for i in range(rows - 2, -1, -1):
            for j in range(cols - 2, -1, -1):
                min_health_needed = min(dp[i + 1][j], dp[i][j + 1])
                dp[i][j] = max(1, min_health_needed - dungeon[i][j])
        return dp[0][0]






#4ms






class Solution:
    def calculateMinimumHP(self, dungeon: List[List[int]]) -> int:
        m, n = len(dungeon), len(dungeon[0])
        INF = float('inf')
        dp = [[INF]*(n+1) for _ in range(m+1)]
        dp[m][n-1] = dp[m-1][n] = 1

        for i in range(m-1, -1, -1):
            for j in range(n-1, -1, -1):
                need = min(dp[i+1][j], dp[i][j+1]) - dungeon[i][j]
                dp[i][j] = 1 if need <= 1 else need

        return dp[0][0]

File: 0101-0200/0174.Dungeon_Game.rs
//0ms





impl Solution {
  // Find the minimum initial health for the knight to be able to rescue the princess.
  pub fn calculate_minimum_hp(dungeon: Vec<Vec<i32>>) -> i32 {
    let m = dungeon.len();
    let n = dungeon[0].len();
    let mut dp = vec![vec![0; n]; m];

    // Start from the princess cell
    dp[m - 1][n - 1] = (1 - dungeon[m - 1][n - 1]).max(1);

    // Fill the last row
    for j in (0..n - 1).rev() {
      dp[m - 1][j] = (dp[m - 1][j + 1] - dungeon[m - 1][j]).max(1);
    }

    // Fill the last column
    for i in (0..m - 1).rev() {
      dp[i][n - 1] = (dp[i + 1][n - 1] - dungeon[i][n - 1]).max(1);
    }

    // Fill the rest of the dp table
    for i in (0..m - 1).rev() {
      for j in (0..n - 1).rev() {
        let min_health_on_exit = dp[i + 1][j].min(dp[i][j + 1]);
        dp[i][j] = (min_health_on_exit - dungeon[i][j]).max(1);
      }
    }

    // The minimum initial health required to reach the princess
    dp[0][0]

  }
}





//0ms







impl Solution {
    pub fn calculate_minimum_hp(dungeon: Vec<Vec<i32>>) -> i32 {
        let m = dungeon.len();
        let n = dungeon[0].len();
        let inf = i32::MAX / 4;
        let mut dp = vec![vec![inf; n + 1]; m + 1];
        dp[m][n - 1] = 1;
        dp[m - 1][n] = 1;

        for i in (0..m).rev() {
            for j in (0..n).rev() {
                let need = dp[i + 1][j].min(dp[i][j + 1]) - dungeon[i][j];
                dp[i][j] = if need <= 1 { 1 } else { need };
            }
        }
        dp[0][0]
    }
}
